File: JavaScript/bucket.js
function insertionSort(bukt) {
    for (let j = 1; j < bukt.length; j++) {
        let val = bukt[j];
        let k = j - 1;
        while (k >= 0 && bukt[k] > val) {
            bukt[k + 1] = bukt[k];
            k--;
        }
        bukt[k + 1] = val;
    }
}

function bucketSort(inputArr) {
    let s = inputArr.length;
    let bucketArr = Array.from({ length: s }, () => []);

    for (let j of inputArr) {
        let bi = Math.floor(s * j);
        if (bi !== s) bucketArr[bi].push(j);
        else bucketArr[s - 1].push(j);
    }

    for (let bukt of bucketArr) insertionSort(bukt);

    let idx = 0;
    for (let bukt of bucketArr)
        for (let j of bukt)
            inputArr[idx++] = j;
}

// Ejemplo
let inputArr = [0.77, 0.16, 0.38, 0.25, 0.71, 0.93, 0.22, 0.11, 0.24, 0.67];
console.log("Arreglo antes de ordenar:");
console.log(inputArr.join(" "));
bucketSort(inputArr);
console.log("Arreglo después de ordenar:");
console.log(inputArr.join(" "));

File: Python/bucket.py
def insertion_sort(bukt):  # Función de ordenamiento por inserción
    for j in range(1, len(bukt)):  # Iterar desde el segundo elemento
        val = bukt[j]              # Elemento actual a insertar
        k = j - 1                  # Índice del elemento anterior
        while k >= 0 and bukt[k] > val:  # Mover elementos mayores hacia la derecha
            bukt[k + 1] = bukt[k]
            k -= 1
        bukt[k + 1] = val          # Insertar elemento en su posición correcta

def bucket_sort(inputArr):  # Función principal de ordenamiento bucket
    s = len(inputArr)       # Número de buckets
    bucketArr = [[] for _ in range(s)]  # Crear lista de buckets vacíos

    # Colocar cada elemento en su bucket correspondiente
    for j in inputArr:
        bi = int(s * j)               # Índice del bucket
        if bi != s:
            bucketArr[bi].append(j)   # Añadir elemento al bucket
        else:
            bucketArr[s - 1].append(j)

    # Ordenar cada bucket individualmente usando ordenamiento por inserción
    for bukt in bucketArr:
        insertion_sort(bukt)

    # Concatenar todos los buckets ordenados en el arreglo original
    idx = 0
    for bukt in bucketArr:
        for j in bukt:
            inputArr[idx] = j
            idx += 1

# Ejemplo de uso
inputArr = [0.77, 0.16, 0.38, 0.25, 0.71, 0.93, 0.22, 0.11, 0.24, 0.67]
print("Arreglo antes de ordenar:")
print(" ".join(map(str, inputArr)))
bucket_sort(inputArr)
print("Arreglo después de ordenar:")
print(" ".join(map(str, inputArr)))